### PR TITLE
Separate out the HW context related changes to different files

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
#### Problem solved by the commit

- Separate out the hw context related code to different file. 
- Data Structure will be separated out to a different file called **kds_hwxtc.h**
- The changes are both XOCL (**xocl_hwctx.c**) and KDS common layer (**kds_hwctx.c**)
- In future HW context related changes will be concentrated more on those new files.
- Existing behavior remains same. No functionality level changes

### CODE REVIEW

**Comment by saifuddin-xilinx** [comment]
retest this please

**Review by chvamshi-xilinx — APPROVED** [review]

**Review by maxzhen — APPROVED** [review]

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_hwctx.c [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+	legacy_args.exec_bo_handle = args->exec_bo_handle;
+
+	return xocl_command_ioctl(xdev, &legacy_args, filp, false);
+}
+
+/*
+ * Open a context (only shared supported today) on a CU under the given hw_context.
+ * Return the acquired cu index for further reference.
+ */
+int xocl_open_cu_ctx_ioctl(struct drm_device *dev, void *data,
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_hwctx.c:273]: All of the ioctl calls are defined in xocl_ioctl.c. Can we keep this rules until we plan to change it?

**Review by mamin506 — CHANGES_REQUESTED** [review]
Just some minor changes.

**Review by stsoe — APPROVED** [review]
Not reviewed in detail.  Apparently I am marked as owner for src/CMake, and those are fine.

File: src/runtime_src/core/common/drv/kds_core.c [hunk 2]
```diff
@@ -45,10 +45,10 @@ int store_kds_echo(struct kds_sched *kds, const char *buf, size_t count,
...
 }
 
-static ssize_t kds_create_cu_string(struct xrt_cu *xcu,
-				    char (*buf)[MAX_CU_STAT_LINE_LENGTH],
-				    int slot, int idx, u64 usage_count,
-				    enum kds_type type)
+ssize_t kds_create_cu_string(struct xrt_cu *xcu,
+		char (*buf)[MAX_CU_STAT_LINE_LENGTH],
+		int slot, int idx, u64 usage_count,
+		enum kds_type type)
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_core.c:51]: This is also get called from kds_populate_cu_buf() which is in kds_core.c. 
this this is not related to hw context, I have placed it here.

File: src/runtime_src/core/common/drv/kds_hwctx.c [hunk 3]
```diff
@@ -0,0 +1,423 @@
+/* SPDX-License-Identifier: GPL-2.0 OR Apache-2.0 */
+/*
+ * Xilinx Kernel Driver Scheduler
+ *
+ * Copyright (C) 2020-2022 Xilinx, Inc. All rights reserved.
+ * Copyright (C) 2022-2023 Advanced Micro Devices, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/drv/kds_hwctx.c:6]: Fixed, Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_hwctx.c [hunk 4]
```diff
@@ -0,0 +1,366 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Xilinx Alveo User Function Driver
+ *
+ * Copyright (C) 2020-2023 Xilinx, Inc.
+ * Copyright (C) 2022-2023 Advanced Micro Devices, Inc.
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_hwctx.c:6]: Fixed. Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_hwctx.c (hunk 1 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_hwctx.c:273]: Sure. Thanks

File: src/runtime_src/core/common/drv/kds_core.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mamin506) [reply src/runtime_src/core/common/drv/kds_core.c:51]: okay, now I see. In show_kds_cuctx_stat_raw(), there are very similar code blocks like kds_populate_cu_buf(). There is space to improve.

But this is for sysfs node, not something must work on right now.

**Review by mamin506 — APPROVED** [review]
Looks good to me.

**Pull request merged**